### PR TITLE
Support `SETTINGS` clause for `S3` engine

### LINKED ISSUE #27580: S3 table engine doesn't support SETTINGS
**Describe the unexpected behaviour**
It should be possible to pass input_format related settings to S3 table engine. (like for URL table engine)

**How to reproduce**
Clickhouse server 21.9

```
CREATE TABLE table_with_range (name String, value UInt32) ENGINE = S3('https://storage.yandexcloud.net/my-test-bucket-768/{some,another}_prefix/some_file_{1..3}', 'CSV') SETTINGS input_format_with_names_use_header=0;

CREATE TABLE table_with_range
(
    `name` String,
    `value` UInt32
)
ENGINE = S3('https://storage.yandexcloud.net/my-test-bucket-768/{some,another}_prefix/some_file_{1..3}', 'CSV')
SETTINGS input_format_with_names_use_header = 0

0 rows in set. Elapsed: 0.011 sec.

Received exception from server (version 21.9.1):
Code: 36. DB::Exception: Received from localhost:9000. DB::Exception: Engine S3 doesn't support SETTINGS clause. Currently only the following engines have support for the feature: [RabbitMQ, Kafka, MySQL, MaterializedPostgreSQL, Join, Set, MergeTree, Memory, URL, ReplicatedVersionedCollapsingMergeTree, ReplacingMergeTree, ReplicatedSummingMergeTree, ReplicatedAggregatingMergeTree, ReplicatedCollapsingMergeTree, File, ReplicatedGraphiteMergeTree, ReplicatedMergeTree, ReplicatedReplacingMergeTree, VersionedCollapsingMergeTree, SummingMergeTree, Distributed, TinyLog, GraphiteMergeTree, CollapsingMergeTree, AggregatingMergeTree, StripeLog, Log]. (BAD_ARGUMENTS)
```

### PR DESCRIPTION
**Pull request by nikitamikhaylov**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category:
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Accept user settings related to file formats in `SETTINGS` clause in `CREATE` query. This closes #27580

### CODE REVIEW

**Comment by nikitamikhaylov** [comment]
The code is pretty the same as in #16591. Need to write better test and discuss a bit what to do with table functions.

**Comment by nikitamikhaylov** [comment]
OMG. `<Warning> AWSClient: ClientConfiguration: Retry Strategy will use the default max attempts.`

File: docker/test/stateless/run.sh [hunk 1]
```diff
@@ -17,6 +17,8 @@ dpkg -i package_folder/clickhouse-test_*.deb
 # install test configs
 /usr/share/clickhouse-test/config/install.sh
 
+./setup_minio.sh
```
[the remark this answers is not in the document]
Comment (nikitamikhaylov) [reply docker/test/stateless/run.sh:20]: Downloading moved to Dockerfile.

File: src/Storages/StorageS3.cpp [hunk 2]
```diff
@@ -732,6 +746,34 @@ void registerStorageS3Impl(const String & name, StorageFactory & factory)
...
+        if (args.storage_def->settings)
+        {
+            FormatFactorySettings user_format_settings;
+
+            // Apply changed settings from global context, but ignore the
+            // unknown ones, because we only have the format settings here.
+            const auto & changes = args.getContext()->getSettingsRef().changes();
+            for (const auto & change : changes)
+            {
+                if (user_format_settings.has(change.name))
```
[the remark this answers is not in the document]
Comment (nikitamikhaylov) [reply src/Storages/StorageS3.cpp:762]: For backward compatibility ?

File: src/Storages/StorageS3.cpp [hunk 3]
```diff
@@ -732,6 +746,34 @@ void registerStorageS3Impl(const String & name, StorageFactory & factory)
...
+                {
+                    user_format_settings.set(change.name, change.value);
+                }
+            }
+
+            // Apply changes from SETTINGS clause, with validation.
+            user_format_settings.applyChanges(args.storage_def->settings->changes);
+            format_settings = getFormatSettings(args.getContext(), user_format_settings);
```
[the remark this answers is not in the document]
Comment (nikitamikhaylov) [reply src/Storages/StorageS3.cpp:770]: I don't clearly understand. But other settings like `max_threads` will be got from context directly. This settings will be passed to Input / Output Formats and only there.

**Comment by nikitamikhaylov** [comment]
@Mergifyio update

File: src/Storages/StorageS3.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (vdimir) [reply src/Storages/StorageS3.cpp:770]: I meant what will happen if we set `max_threads` in `CREATE TABLE ... S3 ... SETTINGS`. I've tested just now, in this case we will got `Unknown setting` error, as expected.

File: src/Storages/StorageS3.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (vdimir) [reply src/Storages/StorageS3.cpp:762]: Ok, I thought that empty `FormatFactorySettings` will return `false` for `has` call, but it returns `true` if setting is a format setting

**Comment by vdimir** [comment]
@nikitamikhaylov I tried to make `setup_minio.sh` script convenient for local run to reproduce tests locally, see my commit https://github.com/ClickHouse/ClickHouse/pull/28037/commits/5a9604ba78407ea6e5449e9bdf46d7332cd49991

Also test fixed (reference files contained spaces instead of tabs): https://github.com/ClickHouse/ClickHouse/pull/28037/commits/088761126138c5048b502cd554d21c8384f8d4b3

**Review by vdimir — APPROVED** [review]

**Comment by nikitamikhaylov** [comment]
@vdimir Thanks a lot!

**Comment by gyuton** [comment]
Internal documentation ticket: DOCSUP-14532.

**Pull request merged**